File: cpp_src/puppy_core/mission_manager.cpp
// Mission Manager 实现
#include "puppy_core/mission_manager.h"

#include <algorithm>
#include <cmath>
#include <sstream>

namespace puppy_core {

MissionManagerNode::MissionManagerNode()
    : rclcpp::Node("mission_manager") {
  // 默认 dock 目标（对应 Python: get_dock_target({'x':0.0,'y':-2.0,'yaw':0.0})）
  dock_target_.name = "dock";
  dock_target_.x = 0.0;
  dock_target_.y = -2.0;
  dock_target_.yaw = 0.0;

  // 默认巡逻路线（对应 Python: get_patrol_route() 在无配置时返回空列表）
  // 当配置加载不可用时使用空列表，保持与Python一致的"no patrol route"行为。

  nav_client_ = rclcpp_action::create_client<NavigateToPose>(
      this, "navigate_to_pose");

  command_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/mission/command", 10,
      std::bind(&MissionManagerNode::onCommand, this, std::placeholders::_1));
  security_sub_ =
      this->create_subscription<puppy_interfaces::msg::SecurityEvent>(
          "/security/event", 10,
          std::bind(&MissionManagerNode::onSecurity, this,
                    std::placeholders::_1));
  battery_sub_ =
      this->create_subscription<puppy_interfaces::msg::BatteryStatus>(
          "/battery_status", 10,
          std::bind(&MissionManagerNode::onBattery, this,
                    std::placeholders::_1));
  capabilities_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/robot/capabilities", 10,
      std::bind(&MissionManagerNode::onCapabilities, this,
                std::placeholders::_1));

  status_pub_ = this->create_publisher<puppy_interfaces::msg::PatrolStatus>(
      "/mission/status", 10);
  timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&MissionManagerNode::publishStatus, this));

  RCLCPP_INFO(this->get_logger(),
              "Mission manager started with patrol route of %zu waypoints",
              patrol_route_.size());
}

void MissionManagerNode::onCapabilities(
    const std_msgs::msg::String::SharedPtr msg) {
  // 保存原始字符串，C++版默认将所有能力视为可用
  // （Python 中 cap is None 时返回 True）
  capabilities_json_ = msg->data;
}

bool MissionManagerNode::isCapabilityAvailable(
    const std::string& /*name*/) const {
  // 简化处理：不解析JSON，保守地将所有能力视为可用
  // 若后续接入完整 JSON 解析，应检查 enabled && available
  return true;
}

void MissionManagerNode::onCommand(
    const std_msgs::msg::String::SharedPtr msg) {
  std::string cmd = msg->data;
  // 转小写
  std::transform(cmd.begin(), cmd.end(), cmd.begin(),
                 [](unsigned char c) { return std::tolower(c); });
  RCLCPP_INFO(this->get_logger(), "Mission command: %s", cmd.c_str());
  if (cmd == "patrol") {
    startPatrol();
  } else if (cmd == "dock") {
    startDocking();
  } else if (cmd == "stop") {
    cancelMission();
  } else if (cmd.rfind("goto:", 0) == 0) {
    startManualNavigation(cmd);
  }
}

void MissionManagerNode::startManualNavigation(const std::string& cmd) {
  if (!isCapabilityAvailable("navigation")) {
    RCLCPP_WARN(this->get_logger(),
                "Navigation capability unavailable - manual goal rejected");
    last_status_message_ = "manual goal rejected: navigation unavailable";
    return;
  }

  // 解析 "goto:x,y"
  double goal_x = 0.0;
  double goal_y = 0.0;
  try {
    auto colon_pos = cmd.find(':');
    if (colon_pos == std::string::npos) {
      throw std::runtime_error("expected x,y");
    }
    std::string coords_str = cmd.substr(colon_pos + 1);
    auto comma_pos = coords_str.find(',');
    if (comma_pos == std::string::npos) {
      throw std::runtime_error("expected x,y");
    }
    std::string x_str = coords_str.substr(0, comma_pos);
    std::string y_str = coords_str.substr(comma_pos + 1);
    // 去除空白
    x_str.erase(std::remove(x_str.begin(), x_str.end(), ' '), x_str.end());
    y_str.erase(std::remove(y_str.begin(), y_str.end(), ' '), y_str.end());
    if (x_str.empty() || y_str.empty()) {
      throw std::runtime_error("expected x,y");
    }
    goal_x = std::stod(x_str);
    goal_y = std::stod(y_str);
  } catch (const std::exception& exc) {
    RCLCPP_WARN(this->get_logger(), "Invalid goto command \"%s\": %s",
                cmd.c_str(), exc.what());
    last_status_message_ = "manual goal rejected: invalid coordinates";
    return;
  }

  if (!nav_client_->wait_for_action_server(std::chrono::seconds(2))) {
    RCLCPP_ERROR(this->get_logger(), "NavigateToPose action not available");
    last_status_message_ =
        "manual goal rejected: navigation action unavailable";
    return;
  }

  cancelMission();
  bool started = startNavigation(
      goal_x, goal_y, 0.0,
      [this](bool success) { this->onGotoResult(success); });
  if (started) {
    current_mission_ = "NAVIGATION";
    std::ostringstream oss;
    oss << "manual goal (" << goal_x << ", " << goal_y << ")";
    last_status_message_ = oss.str();
  }
}

void MissionManagerNode::onSecurity(
    const puppy_interfaces::msg::SecurityEvent::SharedPtr msg) {
  if (msg->severity == "critical") {
    RCLCPP_WARN(this->get_logger(),
                "Security alert: %s - interrupting mission",
                msg->event_type.c_str());
    cancelMission();
  }
}

void MissionManagerNode::onBattery(
    const puppy_interfaces::msg::BatteryStatus::SharedPtr msg) {
  if (msg->low_battery && current_mission_ == "PATROL") {
    RCLCPP_INFO(this->get_logger(), "Low battery - returning to dock");
    cancelMission();
    startDocking();
  }
}

bool MissionManagerNode::startNavigation(double x, double y, double yaw,
                                         CompletionCallback on_complete) {
  if (!nav_client_->wait_for_action_server(std::chrono::seconds(2))) {
    RCLCPP_ERROR(this->get_logger(), "NavigateToPose action not available");
    last_status_message_ = "navigation action unavailable";
    return false;
  }

  NavigateToPose::Goal goal;
  goal.pose.header.frame_id = "map";
  goal.pose.header.stamp = this->now();
  goal.pose.pose.position.x = x;
  goal.pose.pose.position.y = y;
  goal.pose.pose.orientation.z = std::sin(yaw / 2.0);
  goal.pose.pose.orientation.w = std::cos(yaw / 2.0);

  pending_after_nav_ = std::move(on_complete);

  auto send_goal_options =
      rclcpp_action::Client<NavigateToPose>::SendGoalOptions();
  send_goal_options.goal_response_callback =
      std::bind(&MissionManagerNode::onNavGoalResponse, this,
                std::placeholders::_1);
  send_goal_options.result_callback =
      std::bind(&MissionManagerNode::onNavResult, this,
                std::placeholders::_1);
  nav_client_->async_send_goal(goal, send_goal_options);
  return true;
}

void MissionManagerNode::startPatrol() {
  if (patrol_route_.empty()) {
    RCLCPP_ERROR(this->get_logger(), "No patrol route configured");
    last_status_message_ = "no patrol route configured";
    return;
  }
  if (!isCapabilityAvailable("patrol")) {
    RCLCPP_WARN(this->get_logger(),
                "Patrol capability unavailable - patrol rejected");
    current_mission_ = "IDLE";
    last_status_message_ = "patrol rejected: capability unavailable";
    return;
  }
  if (!isCapabilityAvailable("navigation")) {
    RCLCPP_WARN(this->get_logger(),
                "Navigation capability unavailable - patrol rejected");
    current_mission_ = "IDLE";
    last_status_message_ = "patrol rejected: navigation unavailable";
    return;
  }

  cancelMission();
  current_mission_ = "PATROL";
  current_waypoint_index_ = 0;
  last_status_message_ = "patrol started";
  dispatchPatrolWaypoint();
}

void MissionManagerNode::startDocking() {
  if (!isCapabilityAvailable("docking")) {
    RCLCPP_WARN(this->get_logger(),
                "Docking capability unavailable - docking rejected");
    current_mission_ = "IDLE";
    last_status_message_ = "docking rejected: capability unavailable";
    return;
  }
  if (!isCapabilityAvailable("navigation")) {
    RCLCPP_WARN(this->get_logger(),
                "Navigation capability unavailable - docking rejected");
    current_mission_ = "IDLE";
    last_status_message_ = "docking rejected: navigation unavailable";
    return;
  }

  cancelMission();
  current_mission_ = "DOCKING";
  current_waypoint_index_ = -1;
  last_status_message_ = "returning to dock";
  bool started = startNavigation(
      dock_target_.x, dock_target_.y, dock_target_.yaw,
      [this](bool success) { this->onDockResult(success); });
  if (!started) {
    current_mission_ = "IDLE";
  }
}

void MissionManagerNode::cancelMission() {
  if (goal_handle_) {
    try {
      nav_client_->async_cancel_goal(goal_handle_);
    } catch (const std::exception& exc) {
      RCLCPP_WARN(this->get_logger(),
                  "Failed to cancel navigation goal: %s", exc.what());
    }
  }
  goal_handle_.reset();
  pending_after_nav_ = nullptr;
  current_waypoint_index_ = -1;
  current_mission_ = "IDLE";
  last_status_message_ = "mission cancelled";
  RCLCPP_INFO(this->get_logger(), "Mission cancelled");
}

void MissionManagerNode::dispatchPatrolWaypoint() {
  if (current_mission_ != "PATROL") {
    return;
  }

  if (current_waypoint_index_ >= static_cast<int>(patrol_route_.size())) {
    RCLCPP_INFO(this->get_logger(),
                "Patrol route complete - returning to dock");
    startDocking();
    return;
  }

  const Waypoint& waypoint = patrol_route_[current_waypoint_index_];
  std::ostringstream name_oss;
  if (waypoint.name.empty()) {
    name_oss << current_waypoint_index_;
  } else {
    name_oss << waypoint.name;
  }
  last_status_message_ = "navigating to " + name_oss.str();
  RCLCPP_INFO(this->get_logger(),
              "Patrol waypoint %d/%zu: %s (%.2f, %.2f)",
              current_waypoint_index_ + 1, patrol_route_.size(),
              waypoint.name.empty() ? "waypoint" : waypoint.name.c_str(),
              waypoint.x, waypoint.y);
  bool started = startNavigation(
      waypoint.x, waypoint.y, waypoint.yaw,
      [this](bool success) { this->onPatrolWaypointResult(success); });
  if (!started) {
    current_mission_ = "IDLE";
  }
}

void MissionManagerNode::onNavGoalResponse(
    std::shared_future<GoalHandleNavigateToPose::SharedPtr> future) {
  GoalHandleNavigateToPose::SharedPtr goal_handle = future.get();
  if (!goal_handle || !goal_handle->is_active()) {
    // 等价 Python: not goal_handle.accepted
    goal_handle_.reset();
    RCLCPP_ERROR(this->get_logger(), "Navigation goal rejected");
    last_status_message_ = "navigation goal rejected";
    handleNavCompletion(false);
    return;
  }
  goal_handle_ = goal_handle;
}

void MissionManagerNode::onNavResult(
    const GoalHandleNavigateToPose::WrappedResult& result) {
  goal_handle_.reset();
  bool success = (result.code == rclcpp_action::ResultCode::SUCCEEDED);
  if (success) {
    last_status_message_ = "navigation goal reached";
  } else {
    std::ostringstream oss;
    oss << "navigation failed (code=" << static_cast<int>(result.code) << ")";
    last_status_message_ = oss.str();
    RCLCPP_WARN(this->get_logger(), "%s", last_status_message_.c_str());
  }
  handleNavCompletion(success);
}

void MissionManagerNode::handleNavCompletion(bool success) {
  CompletionCallback callback = pending_after_nav_;
  pending_after_nav_ = nullptr;
  if (callback) {
    callback(success);
  }
}

void MissionManagerNode::onPatrolWaypointResult(bool success) {
  if (current_mission_ != "PATROL") {
    return;
  }

  if (!success) {
    const Waypoint& failed = patrol_route_[current_waypoint_index_];
    std::ostringstream name_oss;
    name_oss << (failed.name.empty()
                     ? std::to_string(current_waypoint_index_)
                     : failed.name);
    RCLCPP_WARN(this->get_logger(),
                "Patrol failed at %s; aborting patrol",
                name_oss.str().c_str());
    current_mission_ = "IDLE";
    last_status_message_ = "patrol aborted";
    return;
  }

  const Waypoint& reached = patrol_route_[current_waypoint_index_];
  std::ostringstream name_oss;
  name_oss << (reached.name.empty()
                   ? std::to_string(current_waypoint_index_)
                   : reached.name);
  last_status_message_ = "reached " + name_oss.str();
  current_waypoint_index_ += 1;
  dispatchPatrolWaypoint();
}

void MissionManagerNode::onDockResult(bool success) {
  current_waypoint_index_ = -1;
  current_mission_ = "IDLE";
  if (success) {
    last_status_message_ = "docked";
    RCLCPP_INFO(this->get_logger(), "Dock target reached");
  } else {
    last_status_message_ = "dock navigation failed";
    RCLCPP_WARN(this->get_logger(), "Failed to reach dock target");
  }
}

void MissionManagerNode::onGotoResult(bool success) {
  current_mission_ = "IDLE";
  last_status_message_ = success ? "manual goal reached" : "manual goal failed";
}

void MissionManagerNode::publishStatus() {
  puppy_interfaces::msg::PatrolStatus msg;
  msg.header.stamp = this->now();
  msg.state = current_mission_;
  msg.current_waypoint_index = current_waypoint_index_;
  msg.total_waypoints = static_cast<int32_t>(patrol_route_.size());
  if (current_mission_ == "PATROL" && !patrol_route_.empty()) {
    double ratio = static_cast<double>(current_waypoint_index_) /
                   static_cast<double>(patrol_route_.size());
    msg.completion_ratio = static_cast<float>(std::max(0.0, std::min(1.0, ratio)));
  } else if (current_mission_ == "DOCKING") {
    msg.completion_ratio = 1.0f;
  } else {
    msg.completion_ratio = 0.0f;
  }
  msg.message = last_status_message_;
  status_pub_->publish(msg);
}

}  // namespace puppy_core

File: src/puppy_core/puppy_core/mission_manager.py
"""Mission Manager: high-level task orchestration over Nav2 primitives."""
import json
import math

import rclpy
from action_msgs.msg import GoalStatus
from nav2_msgs.action import NavigateToPose
from rclpy.action import ActionClient
from rclpy.node import Node
from std_msgs.msg import String

from puppy_core.config_utils import get_dock_target, get_patrol_route
from puppy_interfaces.msg import BatteryStatus, PatrolStatus, SecurityEvent


class MissionManagerNode(Node):
    """Orchestrates high-level robot missions."""

    def __init__(self):
        super().__init__('mission_manager')

        self.current_mission = 'IDLE'
        self.dock_target = get_dock_target({'x': 0.0, 'y': -2.0, 'yaw': 0.0})
        self.patrol_route = get_patrol_route()
        self.current_waypoint_index = -1
        self.goal_handle = None
        self.pending_after_nav = None
        self.last_status_message = 'idle'
        self.capabilities = {}

        self.nav_client = ActionClient(self, NavigateToPose, 'navigate_to_pose')

        self.create_subscription(String, '/mission/command', self._on_command, 10)
        self.create_subscription(SecurityEvent, '/security/event', self._on_security, 10)
        self.create_subscription(BatteryStatus, '/battery_status', self._on_battery, 10)
        self.create_subscription(String, '/robot/capabilities', self._on_capabilities, 10)

        self.status_pub = self.create_publisher(PatrolStatus, '/mission/status', 10)
        self.create_timer(0.5, self._publish_status)

        self.get_logger().info(
            f'Mission manager started with patrol route of {len(self.patrol_route)} waypoints'
        )

    def _on_capabilities(self, msg: String):
        try:
            self.capabilities = json.loads(msg.data)
        except Exception:
            pass

    def _is_capability_available(self, name: str) -> bool:
        cap = self.capabilities.get(name)
        if cap is None:
            return True
        return cap.get('enabled', True) and cap.get('available', True)

    def _on_command(self, msg: String):
        cmd = msg.data.lower()
        self.get_logger().info(f'Mission command: {cmd}')
        if cmd == 'patrol':
            self._start_patrol()
        elif cmd == 'dock':
            self._start_docking()
        elif cmd == 'stop':
            self._cancel_mission()
        elif cmd.startswith('goto:'):
            self._start_manual_navigation(cmd)

    def _start_manual_navigation(self, cmd: str):
        if not self._is_capability_available('navigation'):
            self.get_logger().warn('Navigation capability unavailable - manual goal rejected')
            self.last_status_message = 'manual goal rejected: navigation unavailable'
            return

        try:
            parts = cmd.split(':', 1)
            coords = [value.strip() for value in parts[1].split(',')]
            if len(coords) != 2:
                raise ValueError('expected x,y')
            goal_x = float(coords[0])
            goal_y = float(coords[1])
        except (IndexError, ValueError) as exc:
            self.get_logger().warn(f'Invalid goto command "{cmd}": {exc}')
            self.last_status_message = 'manual goal rejected: invalid coordinates'
            return

        if not self.nav_client.wait_for_server(timeout_sec=2.0):
            self.get_logger().error('NavigateToPose action not available')
            self.last_status_message = 'manual goal rejected: navigation action unavailable'
            return

        self._cancel_mission()
        started = self._start_navigation(
            goal_x,
            goal_y,
            0.0,
            self._on_goto_result,
        )
        if started:
            self.current_mission = 'NAVIGATION'
            self.last_status_message = f'manual goal ({goal_x}, {goal_y})'

    def _on_security(self, msg: SecurityEvent):
        if msg.severity == 'critical':
            self.get_logger().warn(f'Security alert: {msg.event_type} - interrupting mission')
            self._cancel_mission()

    def _on_battery(self, msg: BatteryStatus):
        if msg.low_battery and self.current_mission == 'PATROL':
            self.get_logger().info('Low battery - returning to dock')
            self._cancel_mission()
            self._start_docking()

    def _start_navigation(self, x: float, y: float, yaw: float = 0.0, on_complete=None) -> bool:
        if not self.nav_client.wait_for_server(timeout_sec=2.0):
            self.get_logger().error('NavigateToPose action not available')
            self.last_status_message = 'navigation action unavailable'
            return False

        goal = NavigateToPose.Goal()
        goal.pose.header.frame_id = 'map'
        goal.pose.header.stamp = self.get_clock().now().to_msg()
        goal.pose.pose.position.x = float(x)
        goal.pose.pose.position.y = float(y)
        goal.pose.pose.orientation.z = math.sin(float(yaw) / 2.0)
        goal.pose.pose.orientation.w = math.cos(float(yaw) / 2.0)

        self.pending_after_nav = on_complete
        send_future = self.nav_client.send_goal_async(goal)
        send_future.add_done_callback(self._on_nav_goal_response)
        return True

    def _start_patrol(self):
        if not self.patrol_route:
            self.get_logger().error('No patrol route configured')
            self.last_status_message = 'no patrol route configured'
            return
        if not self._is_capability_available('patrol'):
            self.get_logger().warn('Patrol capability unavailable - patrol rejected')
            self.current_mission = 'IDLE'
            self.last_status_message = 'patrol rejected: capability unavailable'
            return
        if not self._is_capability_available('navigation'):
            self.get_logger().warn('Navigation capability unavailable - patrol rejected')
            self.current_mission = 'IDLE'
            self.last_status_message = 'patrol rejected: navigation unavailable'
            return

        self._cancel_mission()
        self.current_mission = 'PATROL'
        self.current_waypoint_index = 0
        self.last_status_message = 'patrol started'
        self._dispatch_patrol_waypoint()

    def _start_docking(self):
        if not self._is_capability_available('docking'):
            self.get_logger().warn('Docking capability unavailable - docking rejected')
            self.current_mission = 'IDLE'
            self.last_status_message = 'docking rejected: capability unavailable'
            return
        if not self._is_capability_available('navigation'):
            self.get_logger().warn('Navigation capability unavailable - docking rejected')
            self.current_mission = 'IDLE'
            self.last_status_message = 'docking rejected: navigation unavailable'
            return

        self._cancel_mission()
        self.current_mission = 'DOCKING'
        self.current_waypoint_index = -1
        self.last_status_message = 'returning to dock'
        started = self._start_navigation(
            self.dock_target['x'],
            self.dock_target['y'],
            self.dock_target.get('yaw', 0.0),
            self._on_dock_result,
        )
        if not started:
            self.current_mission = 'IDLE'

    def _cancel_mission(self):
        if self.goal_handle is not None:
            try:
                self.goal_handle.cancel_goal_async()
            except Exception as exc:
                self.get_logger().warn(f'Failed to cancel navigation goal: {exc}')
        self.goal_handle = None
        self.pending_after_nav = None
        self.current_waypoint_index = -1
        self.current_mission = 'IDLE'
        self.last_status_message = 'mission cancelled'
        self.get_logger().info('Mission cancelled')

    def _dispatch_patrol_waypoint(self):
        if self.current_mission != 'PATROL':
            return

        if self.current_waypoint_index >= len(self.patrol_route):
            self.get_logger().info('Patrol route complete - returning to dock')
            self._start_docking()
            return

        waypoint = self.patrol_route[self.current_waypoint_index]
        self.last_status_message = f'navigating to {waypoint.get("name", self.current_waypoint_index)}'
        self.get_logger().info(
            f'Patrol waypoint {self.current_waypoint_index + 1}/{len(self.patrol_route)}: '
            f'{waypoint.get("name", "waypoint")} ({waypoint["x"]:.2f}, {waypoint["y"]:.2f})'
        )
        started = self._start_navigation(
            waypoint['x'],
            waypoint['y'],
            waypoint.get('yaw', 0.0),
            self._on_patrol_waypoint_result,
        )
        if not started:
            self.current_mission = 'IDLE'

    def _on_nav_goal_response(self, future):
        goal_handle = future.result()
        if goal_handle is None or not goal_handle.accepted:
            self.goal_handle = None
            self.get_logger().error('Navigation goal rejected')
            self.last_status_message = 'navigation goal rejected'
            self._handle_nav_completion(False)
            return

        self.goal_handle = goal_handle
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self._on_nav_result)

    def _on_nav_result(self, future):
        self.goal_handle = None
        try:
            result = future.result()
        except Exception as exc:
            self.get_logger().error(f'Navigation result failed: {exc}')
            self.last_status_message = 'navigation result failed'
            self._handle_nav_completion(False)
            return

        success = result.status == GoalStatus.STATUS_SUCCEEDED
        if success:
            self.last_status_message = 'navigation goal reached'
        else:
            self.last_status_message = f'navigation failed ({result.status})'
            self.get_logger().warn(self.last_status_message)
        self._handle_nav_completion(success)

    def _handle_nav_completion(self, success: bool):
        callback = self.pending_after_nav
        self.pending_after_nav = None
        if callback is not None:
            callback(success)

    def _on_patrol_waypoint_result(self, success: bool):
        if self.current_mission != 'PATROL':
            return

        if not success:
            failed_waypoint = self.patrol_route[self.current_waypoint_index]
            self.get_logger().warn(
                f'Patrol failed at {failed_waypoint.get("name", self.current_waypoint_index)}; aborting patrol'
            )
            self.current_mission = 'IDLE'
            self.last_status_message = 'patrol aborted'
            return

        reached = self.patrol_route[self.current_waypoint_index]
        self.last_status_message = f'reached {reached.get("name", self.current_waypoint_index)}'
        self.current_waypoint_index += 1
        self._dispatch_patrol_waypoint()

    def _on_dock_result(self, success: bool):
        self.current_waypoint_index = -1
        self.current_mission = 'IDLE'
        if success:
            self.last_status_message = 'docked'
            self.get_logger().info('Dock target reached')
        else:
            self.last_status_message = 'dock navigation failed'
            self.get_logger().warn('Failed to reach dock target')

    def _on_goto_result(self, success: bool):
        self.current_mission = 'IDLE'
        self.last_status_message = 'manual goal reached' if success else 'manual goal failed'

    def _publish_status(self):
        msg = PatrolStatus()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.state = self.current_mission
        msg.current_waypoint_index = self.current_waypoint_index
        msg.total_waypoints = len(self.patrol_route)
        if self.current_mission == 'PATROL' and self.patrol_route:
            msg.completion_ratio = min(
                1.0,
                max(0.0, float(self.current_waypoint_index) / float(len(self.patrol_route))),
            )
        elif self.current_mission == 'DOCKING':
            msg.completion_ratio = 1.0
        else:
            msg.completion_ratio = 0.0
        msg.message = self.last_status_message
        self.status_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = MissionManagerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
